schemas/dataSources.schema.json

```diff
@@ -0,0 +1,52 @@
+{
+    "$schema": "http://json-schema.org/draft-04/schema",
+    "id": "https://immersive-web/gamepad-mapping/0.1.0/dataSources.schema.json",
+    "title": "The data sources array",
+    "type": "array",
+    "description": "The data sources array",
+    "additionalItems": false,
+    "uniqueItems": true,
+    "minItems": 1,
+    "items": {
+        "anyOf": [
+            { 
+                "description": "Button data source",
+                "type": "object",
+                "additionalProperties": false,
+                "required": ["id", "dataSourceType", "button"],
+                "properties":{
+                    "id": { "allOf": [{ "$ref": "dataSources.properties.schema.json#/definitions/id" }] },
+                    "dataSourceType": { "enum": ["buttonSource"] },
+                    "button": { "allOf": [{ "$ref": "dataSources.properties.schema.json#/definitions/button" }] }
+                }
+            },
+            { 
+                "description": "Dpad data source",
+                "type": "object",
+                "additionalProperties": false,
+                "required": ["id", "dataSourceType", "left", "right", "down", "up"],
+                "properties":{
+                    "id": { "allOf": [{ "$ref": "dataSources.properties.schema.json#/definitions/id" }] },
+                    "dataSourceType": { "enum": ["dpadSource"] },
+                    "left": { "allOf": [{ "$ref": "dataSources.properties.schema.json#/definitions/button" }] },
+                    "right": { "allOf": [{ "$ref": "dataSources.properties.schema.json#/definitions/button" }] },
+                    "down": { "allOf": [{ "$ref": "dataSources.properties.schema.json#/definitions/button" }] },
+                    "up": { "allOf": [{ "$ref": "dataSources.properties.schema.json#/definitions/button" }] }
```

Comment:
In my experience with traditional gamepads it's not uncommon for Dpads to be exposed as "digital axes", in that they're axis data that only reports -1, 0, or 1. If a UA was reporting device values directly rather than attempting some form of internal mapping you may end up needing to expose a pair of axes as a semantic Dpad.

Reply:
For real?  Ugh.  I was patterning off of the gamepad spec which says they should be individual buttons....

Reply:
Agreed that the browser should really be exposing them as button elements when it can, but in traditional Gamepad land we would run into a lot of cases where a device was exposed as a generic gamepad (not through, say, XInput or a similar well-structured API) and the best way to surface them was just to directly report the axes and buttons as they came off the device HID reports. This sometimes results in DPads that were reported as axes.

In general I expect this to be less of a problem with WebXR, because we have fewer possible sources of controller data and they are generally strictly typed, but it's not impossible that some UAs (or, more likely, custom branches of a UA) would do something that results in less rigorous UA-side mapping.

Reply:
modify the schema to allow for either xAxis/yAxis or 4 buttons in a dpad